File: backend/ros2_visual_backend/safety_logic.py
from dataclasses import dataclass
from math import isfinite

from .config import SafetyConfig


@dataclass(frozen=True)
class SafetyResult:
    linear_x: float
    angular_z: float
    stopped: bool
    front_distance: float
    reason: str


def nearest_front_distance(
    ranges: list[float],
    angle_min: float,
    angle_increment: float,
    range_min: float,
    range_max: float,
    front_angle_deg: float = 15.0,
) -> float:
    limit = front_angle_deg / 180.0 * 3.141592653589793
    values = []
    for index, value in enumerate(ranges):
        angle = angle_min + angle_increment * index
        if abs(angle) <= limit and isfinite(value) and range_min <= value <= range_max:
            values.append(value)
    return min(values, default=float("inf"))


def decide_safe_command(
    linear_x: float,
    angular_z: float,
    ranges: list[float] | None,
    angle_min: float,
    angle_increment: float,
    range_min: float,
    range_max: float,
    now_ms: float,
    last_scan_ms: float,
    last_command_ms: float,
    was_stopped: bool,
    config: SafetyConfig = SafetyConfig(),
) -> SafetyResult:
    if now_ms - last_command_ms > config.command_timeout_sec * 1000:
        return SafetyResult(0.0, 0.0, True, float("inf"), "command-timeout")
    if ranges is None or now_ms - last_scan_ms > config.scan_timeout_sec * 1000:
        return SafetyResult(0.0, 0.0, True, float("inf"), "scan-timeout")
    distance = nearest_front_distance(ranges, angle_min, angle_increment, range_min, range_max, config.front_angle_deg)
    blocked = distance < config.stop_distance
    can_resume = distance >= config.resume_distance or distance == float("inf")
    stopped = linear_x > 0 and (blocked or (was_stopped and not can_resume))
    return SafetyResult(0.0 if stopped else linear_x, angular_z, stopped, distance, "obstacle" if stopped else "clear")

File: src/safetyLogic.ts
import { ControlInput, LaserScanMessage, SafetyDecision, makeTwist, zeroTwist } from './types';

export interface SafetyConfig {
  stopDistance: number;
  resumeDistance: number;
  frontAngleDeg: number;
  scanTimeoutSec: number;
  commandTimeoutSec: number;
}

export const DEFAULT_SAFETY_CONFIG: SafetyConfig = {
  stopDistance: 0.34,
  resumeDistance: 0.42,
  frontAngleDeg: 15,
  scanTimeoutSec: 0.5,
  commandTimeoutSec: 0.5,
};

export function frontDistance(scan: LaserScanMessage, frontAngleDeg = DEFAULT_SAFETY_CONFIG.frontAngleDeg): number {
  const limit = (frontAngleDeg * Math.PI) / 180;
  let closest = Number.POSITIVE_INFINITY;
  scan.ranges.forEach((range, index) => {
    const angle = scan.angle_min + scan.angle_increment * index;
    if (Math.abs(angle) <= limit && Number.isFinite(range) && range >= scan.range_min && range <= scan.range_max) {
      closest = Math.min(closest, range);
    }
  });
  return closest;
}

export function decideSafeCommand(
  raw: ControlInput,
  scan: LaserScanMessage | null,
  now: number,
  lastScanAt: number,
  lastCommandAt: number,
  wasStopped: boolean,
  config: SafetyConfig = DEFAULT_SAFETY_CONFIG,
): SafetyDecision {
  if (now - lastCommandAt > config.commandTimeoutSec * 1000) {
    return { command: zeroTwist(), stopped: true, frontDistance: Number.POSITIVE_INFINITY, reason: 'command-timeout' };
  }
  if (!scan || now - lastScanAt > config.scanTimeoutSec * 1000) {
    return { command: zeroTwist(), stopped: true, frontDistance: Number.POSITIVE_INFINITY, reason: 'scan-timeout' };
  }

  const distance = frontDistance(scan, config.frontAngleDeg);
  const blocked = distance < config.stopDistance;
  const canResume = distance >= config.resumeDistance || !Number.isFinite(distance);
  const stopped = raw.linear > 0 && (blocked || (wasStopped && !canResume));
  return {
    command: stopped ? makeTwist(0, raw.angular) : makeTwist(raw.linear, raw.angular),
    stopped,
    frontDistance: distance,
    reason: stopped ? 'obstacle' : 'clear',
  };
}
